server/src/main/java/org/elasticsearch/indices/recovery/MultiFileReader.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.indices.recovery;
+
+import org.apache.lucene.index.CorruptIndexException;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.lucene.store.InputStreamIndexInput;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.core.internal.io.IOUtils;
+import org.elasticsearch.index.store.Store;
+import org.elasticsearch.index.store.StoreFileMetaData;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.Objects;
+
+/**
+ * A thread safe implementation that allows reading multiple files sequentially chunk by chunk.
+ */
+final class MultiFileReader implements Closeable {
+    private boolean closed = false;
+    private final Store store;
+    private final Iterator<StoreFileMetaData> remainingFiles;
+    private StoreFileMetaData md;
+    private InputStreamIndexInput currentInput = null;
+    private long currentChunkPosition = 0;
+    private final int chunkSizeInBytes;
+    private final Deque<byte[]> recycledBuffers;
+
+    MultiFileReader(Store store, StoreFileMetaData[] files, int chunkSizeInBytes) {
+        this.store = store;
+        this.remainingFiles = Arrays.asList(files).iterator();
+        this.chunkSizeInBytes = chunkSizeInBytes;
+        this.recycledBuffers = ConcurrentCollections.newDeque();
+    }
+
+    synchronized StoreFileMetaData currentFile() {
+        assert md != null;
+        return md;
+    }
+
+    /**
+     * Reads the next file chunk if available. This method returns {@code null} when all provided files are exhaustedly read.
+     * When the caller no longer needs the returned file chunk, it should call {@link FileChunk#close()} to release that file
+     * chunk so the reader can reuse the associated buffer to reduce allocation.
+     */
+    synchronized FileChunk readNextChunk() throws IOException {
```

Comment:
Question: Shouldn't we in fact not need any of the synchronization here? It seems to me that the `Semaphore` in `org.elasticsearch.indices.recovery.RecoverySourceHandler.MultiFileSender` should ensure that we have all the locking we need here?

On a related note (and very optional): Maybe it would also make sense to move this class into `org.elasticsearch.indices.recovery.RecoverySourceHandler`  since it's only used there and to make this kind of relationship a little easier to read?

Reply:
+1. I pushed https://github.com/elastic/elasticsearch/pull/44040/commits/5c1977a2bf8e6f29c4f79ae5d8eae1d5162611a2.